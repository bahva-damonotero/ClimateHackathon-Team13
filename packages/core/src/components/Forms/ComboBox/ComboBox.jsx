import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash.isequal';
import comboBox from '@uswds/uswds/js/usa-combo-box';

import { FormError, Label } from '..';

export default function ComboBox({
  children,
  comparator,
  disabled,
  error,
  filter,
  id,
  onChange,
  options,
  optionRenderer,
  placeholder,
  value,
  ...props
}) {
  const containerRef = useRef();
  const comboBoxRef = useRef();
  const selectRef = useRef();
  const valueRef = useRef(value);

  const pristineClass = 'usa-combo-box--pristine';

  const selectionIndex = (newValue) => {
    const index = options.findIndex((option) =>
      comparator ? comparator(option, newValue) : isEqual(option, newValue)
    );

    return index >= 0 ? index : undefined;
  };

  // initialization
  useEffect(() => {
    const containerEl = containerRef.current;
    const selectEl = selectRef.current;
    if (!containerEl || !selectEl) return;

    // call USWDS's `on` function to initialize the component
    comboBox.on(containerEl);

    // hook into the select's `change` event to monitor changes
    selectEl.addEventListener('change', (e) => {
      if (!onChange) return;

      // do not trigger onChange if the value did not change
      if (
        selectionIndex(valueRef.current) === Number.parseInt(e.detail.value, 10)
      ) {
        return;
      }

      const newValue = options[e.detail.value];
      valueRef.current = newValue;

      const event = new CustomEvent('change', {
        detail: {
          name: id,
          value: newValue
        }
      });
      onChange(event);
    });

    // on dismount, call USWDS's `off` function to cleanup the component
    return () => {
      comboBox.off(containerEl);
    };
  }, []);

  // update the combo box's value whenever the value changes outside
  // of this component
  useEffect(() => {
    const comboBoxEl = comboBoxRef.current;
    const selectEl = selectRef.current;
    if (!comboBoxEl || !selectRef) return;

    const inputEl = comboBoxEl.querySelector('input');

    valueRef.current = value;
    const optionIndex = selectionIndex(value);
    let valueText = '';

    if (optionIndex !== undefined) {
      const valueOption = selectEl.querySelector(
        `option[value="${optionIndex}"]`
      );
      valueText = valueOption.innerHTML;
    }

    if (valueText !== inputEl.value) {
      inputEl.value = valueText;

      if (value) {
        selectEl.value = optionIndex;
        comboBoxEl.classList.add(pristineClass);
      } else {
        // clear the combo box
        selectEl.value = '';
        comboBoxEl.classList.remove(pristineClass);
      }
    }
  }, [value, comboBoxRef, selectRef]);

  // call USWDS's `enable` or `disable` function when the disabled prop changes
  useEffect(() => {
    const comboBoxEl = comboBoxRef.current;
    if (!comboBoxEl) return;

    const disabledState = comboBoxEl.querySelector(`#${id}`).disabled;
    if (disabled === disabledState) return; // the input already has the right disabled state

    if (disabled) comboBox.disable(comboBoxEl);
    else comboBox.enable(comboBoxEl);
  }, [disabled]);

  return (
    <div className="bah-uswds-combo-box" ref={containerRef}>
      <Label error={error} target={id}>
        {children}
      </Label>

      <FormError error={error} />

      <div
        className="usa-combo-box"
        data-default-value={selectionIndex(value)}
        data-filter={filter}
        data-placeholder={placeholder}
        ref={comboBoxRef}
      >
        <select
          className="usa-select usa-combo-box__select"
          disabled={disabled}
          id={id}
          name={id}
          ref={selectRef}
          {...props}
        >
          {options.map((option, index) => (
            <option key={`${id}--${index}`} value={index}>
              {optionRenderer ? optionRenderer(option) : option}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
}

ComboBox.propTypes = {
  /** Label for input */
  children: PropTypes.node.isRequired,
  /**
   * Let's you specify a custom function for comparing the selected value with the list of options.
   * If this is not provided, lodash's isEqual function is used: https://lodash.com/docs/#isEqual.
   */
  comparator: PropTypes.func,
  /** Should the combo box be disabled? */
  disabled: PropTypes.bool,
  /** Error message to display */
  error: PropTypes.string,
  /**
   * The combo box will use this regular expression to filter the combo box options.
   * You are declaring a case insensitive match over the entire option text, which
   * means ^ and $ are added automatically. You can specify the inputted query with {{query}}.
   */
  filter: PropTypes.string,
  /** ID of the input */
  id: PropTypes.string.isRequired,
  /** Function called whenever the combo box selection is changed */
  onChange: PropTypes.func,
  /** Specifies how the options should be displayed in the combo box */
  optionRenderer: PropTypes.func,
  /** Array of combo box options */
  options: PropTypes.array.isRequired,
  /** Text to show when a value has not been selected */
  placeholder: PropTypes.string,
  /** Used to set the combo box selection */
  value: PropTypes.any
};

ComboBox.defaultProps = {
  disabled: false
};
