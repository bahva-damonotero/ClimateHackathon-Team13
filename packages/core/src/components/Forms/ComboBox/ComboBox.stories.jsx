import React from 'react';
import { action } from '@storybook/addon-actions';

import ComboBox from '.';
import { Form } from '..';

export default {
  title: 'Forms/ComboBox',
  component: ComboBox
};

const Template = (args) => (
  <Form>
    <ComboBox {...args} />
  </Form>
);

const defaultArgs = {
  children: 'Select your desired option',
  id: 'combobox-example',
  options: [
    { firstName: 'John', lastName: 'Doe' },
    { firstName: 'Jane', lastName: 'Doe' },
    { firstName: 'Bob', lastName: 'Bobson' }
  ],
  optionRenderer: ({ firstName, lastName }) => `${firstName} ${lastName}`,
  onChange: action('onChange')
};

export const def = Template.bind({});
def.storyName = 'Default';
def.args = {
  ...defaultArgs
};

export const selection = Template.bind({});
selection.args = {
  ...defaultArgs,
  value: defaultArgs.options[0]
};

export const customComparator = Template.bind({});
customComparator.args = {
  ...defaultArgs,
  comparator: (value, option) =>
    value && option
      ? value.firstName === option.firstName &&
        value.lastName === option.lastName
      : false,
  value: defaultArgs.options[2]
};

export const placeholder = Template.bind({});
placeholder.args = {
  ...defaultArgs,
  placeholder: 'Select an option'
};

export const error = Template.bind({});
error.args = {
  ...defaultArgs,
  error: 'Invalid selection'
};

export const stringArray = Template.bind({});
stringArray.args = {
  ...defaultArgs,
  options: ['Bob Bobson', 'Jane Doe', 'John Doe'],
  optionRenderer: undefined
};
