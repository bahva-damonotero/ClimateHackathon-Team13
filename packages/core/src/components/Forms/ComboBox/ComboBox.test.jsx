/* eslint-disable import/first */
jest.mock('lodash.isequal');

import React from 'react';
import { createEvent, fireEvent, render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import comboBox from '@uswds/uswds/js/usa-combo-box';
import isEqual from 'lodash.isequal';

import ComboBox from '.';

describe('ComboBox', () => {
  const id = 'combobox-example';
  const label = 'Select your desired option';

  const basicOptions = ['John Doe', 'Jane Doe', 'Bob Bobson'];

  const options = [
    { firstName: 'John', lastName: 'Doe' },
    { firstName: 'Jane', lastName: 'Doe' },
    { firstName: 'Bob', lastName: 'Bobson' }
  ];

  const optionRenderer = ({ firstName, lastName }) =>
    `${firstName} ${lastName}`;

  const comparator = (value, option) =>
    value && option
      ? value.firstName === option.firstName &&
        value.lastName === option.lastName
      : false;

  const INPUT_SELECTOR = `#${id}.usa-combo-box__input`;
  const SELECT_SELECTOR = `.usa-combo-box__select`;
  const LABEL_SELECTOR = `#${id}-label.usa-label`;
  const PRISTINE_SELECTOR = '.usa-combo-box--pristine';

  it('should render correctly', () => {
    const { container } = render(
      <ComboBox id={id} options={basicOptions}>
        {label}
      </ComboBox>
    );

    expect(container.querySelector(INPUT_SELECTOR)).toBeTruthy();

    const labelNode = container.querySelector(LABEL_SELECTOR);
    expect(labelNode.innerHTML).toBe(label);
    expect(labelNode.getAttribute('for')).toBe(id);
  });

  it('should call `comboBox.on()` to initialize the component', () => {
    const onSpy = jest.spyOn(comboBox, 'on');

    render(
      <ComboBox id={id} options={basicOptions}>
        {label}
      </ComboBox>
    );

    expect(onSpy).toHaveBeenCalled();
  });

  it('should call `onChange` when the value changes', () => {
    const onChange = jest.fn();

    const { container } = render(
      <ComboBox
        comparator={comparator}
        id={id}
        onChange={onChange}
        optionRenderer={optionRenderer}
        options={options}
        value={options[1]}
      >
        {label}
      </ComboBox>
    );

    const selectNode = container.querySelector(SELECT_SELECTOR);

    fireEvent(
      selectNode,
      createEvent(
        'change',
        selectNode,
        {
          detail: {
            value: 0
          }
        },
        { EventType: 'CustomEvent' }
      )
    );

    expect(onChange).toHaveBeenCalledWith(
      new CustomEvent('change', {
        detail: {
          name: id,
          value: options[0]
        }
      })
    );
  });

  it('should not call `onChange` when the value stays the same', () => {
    const onChange = jest.fn();

    const { container } = render(
      <ComboBox
        comparator={comparator}
        id={id}
        onChange={onChange}
        optionRenderer={optionRenderer}
        options={options}
        value={options[1]}
      >
        {label}
      </ComboBox>
    );

    const selectNode = container.querySelector(SELECT_SELECTOR);

    fireEvent(
      selectNode,
      createEvent(
        'change',
        selectNode,
        {
          detail: {
            value: 1
          }
        },
        { EventType: 'CustomEvent' }
      )
    );

    expect(onChange).not.toHaveBeenCalled();
  });

  it('should be able to update the value programmatically', () => {
    // render the combo box with no value
    const initialProps = {
      comparator,
      id,
      onChange: () => {},
      optionRenderer,
      options
    };

    const { container, rerender } = render(
      <ComboBox {...initialProps}>{label}</ComboBox>
    );

    expect(container.querySelector(PRISTINE_SELECTOR)).toBeFalsy();

    const inputNode = container.querySelector(INPUT_SELECTOR);
    const selectNode = container.querySelector(SELECT_SELECTOR);

    expect(inputNode.value).toBe('');
    expect(selectNode.value).toBe('');

    // update the combo box's value programmatically/externally
    const newValueIndex = 0;
    const newValue = options[newValueIndex];

    rerender(
      <ComboBox {...initialProps} value={newValue}>
        {label}
      </ComboBox>
    );

    expect(inputNode.value).toBe(optionRenderer(newValue));
    expect(selectNode.value).toBe(`${newValueIndex}`);

    expect(container.querySelector(PRISTINE_SELECTOR)).toBeTruthy();
  });

  it('should be able to clear the value programmatically', () => {
    // render the combo box with a value
    const initialValueIndex = 0;
    const initialValue = options[initialValueIndex];

    const initialProps = {
      comparator,
      id,
      onChange: () => {},
      optionRenderer,
      options,
      value: initialValue
    };

    const { container, rerender } = render(
      <ComboBox {...initialProps}>{label}</ComboBox>
    );

    expect(container.querySelector(PRISTINE_SELECTOR)).toBeTruthy();

    const inputNode = container.querySelector(INPUT_SELECTOR);
    const selectNode = container.querySelector(SELECT_SELECTOR);

    expect(inputNode.value).toBe(optionRenderer(initialValue));
    expect(selectNode.value).toBe(`${initialValueIndex}`);

    // clear the combo box's value programmatically/externally
    rerender(
      <ComboBox {...initialProps} value={undefined}>
        {label}
      </ComboBox>
    );

    expect(inputNode.value).toBe('');
    expect(selectNode.value).toBe('');

    expect(container.querySelector(PRISTINE_SELECTOR)).toBeFalsy();
  });

  it('should use `lodash.isEqual` when `comparator` is not provided', () => {
    // render the combo box with no value
    const initialProps = {
      id,
      onChange: () => {},
      optionRenderer,
      options
    };

    const { rerender } = render(<ComboBox {...initialProps}>{label}</ComboBox>);

    // update the combo box's value programmatically/externally
    const newValue = options[0];

    rerender(
      <ComboBox {...initialProps} value={newValue}>
        {label}
      </ComboBox>
    );

    expect(isEqual).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to true should call `comboBox.disable()', () => {
    const disableSpy = jest.spyOn(comboBox, 'disable');

    // render the combo box enabled
    const initialProps = {
      id,
      onChange: () => {},
      options: basicOptions
    };

    const { rerender } = render(<ComboBox {...initialProps}>{label}</ComboBox>);

    // disable the combo box
    rerender(
      <ComboBox {...initialProps} disabled>
        {label}
      </ComboBox>
    );

    expect(disableSpy).toHaveBeenCalled();
  });

  it('changing the `disabled` prop to false should call `comboBox.enable()', () => {
    const enableSpy = jest.spyOn(comboBox, 'enable');

    // render the combo box disabled
    const sharedProps = {
      id,
      onChange: () => {},
      options: basicOptions
    };

    const { rerender } = render(
      <ComboBox {...sharedProps} disabled>
        {label}
      </ComboBox>
    );

    // enable the combo box
    rerender(<ComboBox {...sharedProps}>{label}</ComboBox>);

    expect(enableSpy).toHaveBeenCalled();
  });

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <ComboBox id={id} options={basicOptions}>
          {label}
        </ComboBox>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
