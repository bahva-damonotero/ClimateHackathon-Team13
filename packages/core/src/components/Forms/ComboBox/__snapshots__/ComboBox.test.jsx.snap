// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ComboBox should match snapshot 1`] = `
<div
  className="bah-uswds-combo-box"
>
  <label
    className="usa-label"
    htmlFor="combobox-example"
  >
    Select your desired option
  </label>
  <div
    className="usa-combo-box"
  >
    <select
      className="usa-select usa-combo-box__select"
      disabled={false}
      id="combobox-example"
      name="combobox-example"
    >
      <option
        value={0}
      >
        John Doe
      </option>
      <option
        value={1}
      >
        Jane Doe
      </option>
      <option
        value={2}
      >
        Bob Bobson
      </option>
    </select>
  </div>
</div>
`;
